File: lingting-spring-elasticsearch/src/main/java/live/lingting/spring/elasticsearch/ElasticsearchServiceImplBeanPost.java
package live.lingting.spring.elasticsearch;

import co.elastic.clients.elasticsearch.ElasticsearchClient;
import live.lingting.framework.elasticsearch.ElasticsearchApi;
import live.lingting.framework.elasticsearch.ElasticsearchProperties;
import live.lingting.framework.elasticsearch.ElasticsearchProvider;
import live.lingting.framework.elasticsearch.datascope.ElasticsearchDataPermissionHandler;
import live.lingting.spring.post.SpringBeanPostProcessor;
import lombok.RequiredArgsConstructor;

/**
 * @author lingting 2024-03-08 16:52
 */
@RequiredArgsConstructor
@SuppressWarnings("java:S3740")
public class ElasticsearchServiceImplBeanPost implements SpringBeanPostProcessor {

	private final ElasticsearchProperties properties;

	private final ElasticsearchClient client;

	private final ElasticsearchDataPermissionHandler handler;

	@Override
	public boolean isProcess(Object bean, String beanName, boolean isBefore) {
		return bean instanceof AbstractElasticsearchServiceImpl<?>;
	}

	@Override
	@SuppressWarnings({ "unchecked" })
	public Object postProcessAfter(Object bean, String beanName) {
		if (bean instanceof AbstractElasticsearchServiceImpl impl) {
			ElasticsearchApi api = ElasticsearchProvider.api(impl.index, impl.cls, impl::documentId, properties,
					handler, client);
			impl.setApi(api);
		}
		return bean;
	}

}

File: lingting-spring-elasticsearch/src/main/kotlin/live/lingting/spring/elasticsearch/ElasticsearchServiceImplBeanPost.kt
package live.lingting.spring.elasticsearch

import co.elastic.clients.elasticsearch.ElasticsearchClient
import live.lingting.framework.elasticsearch.ElasticsearchProperties
import live.lingting.framework.elasticsearch.api.ElasticsearchApiImpl
import live.lingting.framework.elasticsearch.interceptor.Interceptor
import live.lingting.framework.elasticsearch.polymerize.PolymerizeFactory
import live.lingting.spring.post.SpringBeanPostProcessor

/**
 * @author lingting 2024-03-08 16:52
 */
class ElasticsearchServiceImplBeanPost(
    val properties: ElasticsearchProperties,
    val client: ElasticsearchClient,
    val interceptors: List<Interceptor>,
    val polymerizeFactory: PolymerizeFactory,
) : SpringBeanPostProcessor {

    override fun isProcess(bean: Any, beanName: String, isBefore: Boolean): Boolean {
        return bean is AbstractElasticsearchServiceImpl<*>
    }

    @Suppress("UNCHECKED_CAST")
    override fun postProcessAfter(bean: Any, beanName: String): Any {
        if (bean is AbstractElasticsearchServiceImpl<*>) {
            val impl = (bean as AbstractElasticsearchServiceImpl<Any>)
            val api = ElasticsearchApiImpl(impl.cls, polymerizeFactory, { impl.documentId(it) }, properties, interceptors, client)
            impl.api = api
        }
        return bean
    }
}
